datalog/datalog.go

```diff
@@ -498,368 +269,66 @@ func (m MatchedVariables) Clone() MatchedVariables {
-}
 
-func (c *Combinator) Combine(syms *SymbolTable) ([]map[Variable]*Term, error) {
-	var variables []map[Variable]*Term
-
-	// Stop when no more predicates are available
-	if len(c.predicates) == 0 {
-		if vars := c.variables.Complete(); vars != nil {
-			valid := true
-			for _, e := range c.expressions {
-				res, err := e.Evaluate(vars, syms)
-				if err != nil {
-					return nil, err
-				}
-				if !res.Equal(Bool(true)) {
-					valid = false
-					break
-				}
-			}
-
-			if valid {
-				variables = append(variables, vars)
+	solv, err := NewSolver(r.Body, r.Expressions, facts, syms)
+	if err != nil {
+		return err
+	}
+	err = solv.Solve(0, variables)
+	if err != nil {
+		return err
+	}
+	if len(solv.solutions) == 0 {
+		switch len(variables) {
+		case 0:
+			// special case, query with Facts
+			return nil
+		default:
+			myErr := &SolvSafeForWorldError{
+				errMsg: "datalog: solver didn't find a matching solution",
 			}
+			return myErr
 		}
-		return variables, nil
 	}
-
-	for i, pred := range c.predicates {
-		for ii, currentFact := range *c.currentFacts {
-			vars := c.variables.Clone()
-			matchIDs := true
-			// minLen is the smallest number of IDs
-			// between the predicate and the current fact
-			minLen := len(pred.Terms)
-			if l := len(currentFact.Predicate.Terms); l < minLen {
-				minLen = l
-			}
-
-			for j := 0; j < minLen; j++ {
-				id := pred.Terms[j]
-				k, ok := id.(Variable)
-				if !ok {
-					continue
-				}
-				v := currentFact.Predicate.Terms[j]
-				if !vars.Insert(k, v) {
-					matchIDs = false
-				}
-				if !matchIDs {
-					break
-				}
-			}
-
-			if !matchIDs {
+outer:
+	for _, h := range solv.solutions {
+		p := r.Head.Clone()
+		for i, id := range p.Terms {
+			k, ok := id.(Variable)
+			if !ok {
+				// TODO: Check if this really works with a Rule where head has a non-var Pred
+				// see test13, test Token has the follwing in block 1
+				// valid_date("file1") <- time($0), resource("file1"), $0 <= 2030-12-31T12:59:59Z;
 				continue
 			}
+			v, ok := h[k]
+			if !ok {
+				return InvalidRuleError{r, k}
+			}
 
-			if len(c.predicates) > i+1 {
-				next, err := NewCombinator(vars, c.predicates[i+1:], c.expressions, c.allFacts).Combine(syms)
-				if err != nil {
-					return nil, err
-				}
-				if len(next) == 0 {
-					// returns only if there is no more current facts, otherwise process next one
-					if ii == len(*c.currentFacts)-1 {
-						return variables, nil
-					}
-					continue
-				}
-				variables = append(variables, next...)
-			} else {
-				if v := vars.Complete(); v != nil {
-					valid := true
-					for _, e := range c.expressions {
-						res, err := e.Evaluate(v, syms)
-						if err != nil {
-							return nil, err
-						}
-						if !res.Equal(Bool(true)) {
-							valid = false
-							break
-						}
-					}
-
-					if valid {
-						variables = append(variables, v)
-					}
+			// prevent the rule from generating facts with forbidden IDs
+			for _, f := range r.forbiddenIDs {
```

Comment:
this can be removed now, the forbidden ids were only needed in 1.0 (I forgot to remove it in 2.0)